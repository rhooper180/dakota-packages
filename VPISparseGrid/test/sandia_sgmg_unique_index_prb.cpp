# include "sandia_rules.hpp"
# include "sandia_rules2.hpp"
# include "sandia_sgmg.hpp"

# include <cstdlib>
# include <iomanip>
# include <iostream>
# include <cmath>
//
//  Two global variables needed to support the "parameter" function.
//
double *P;
int *NP;

int main ( );

void sgmg_unique_index_tests ( double tol );

void sgmg_unique_index_test 
(
  int dim_num,
  int level_max_min, 
  int level_max_max,
  int growth, 
  void ( *gw_compute_points[] ) ( int order, int dim, double w[] ),
  double tol,
  int ( *gw_compute_order[] ) ( int level, int growth ) 
);

typedef void ( *GWPointer ) ( int order, int dim, double w[] );
typedef int ( *GWPointer2 ) ( int level, int growth );

//****************************************************************************80

int main ( )

//****************************************************************************80
//
//  Purpose:
//
//    MAIN tests SGMG_UNIQUE_INDEX.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license.
//
//  Modified:
//
//    01 January 2012
//
//  Author:
//
//    John Burkardt
//
//  Reference:
//
//    Fabio Nobile, Raul Tempone, Clayton Webster,
//    A Sparse Grid Stochastic Collocation Method for Partial Differential
//    Equations with Random Input Data,
//    SIAM Journal on Numerical Analysis,
//    Volume 46, Number 5, 2008, pages 2309-2345.
//
{
  double tol;

  webbur::timestamp ( );
  std::cout << "\n";
  std::cout << "SGMG_UNIQUE_INDEX_PRB\n";
  std::cout << "  C++ version\n";
//
//  Check that we can determine a UNIQUE_INDEX mapping, so that we can
//  generate all the points, and then select unique representatives, and
//  then match each point to its representative.
//
  tol = std::sqrt ( webbur::r8_epsilon ( ) );
  sgmg_unique_index_tests ( tol );
//
//  Terminate.
//
  std::cout << "\n";
  std::cout << "SGMG_UNIQUE_INDEX_PRB\n";
  std::cout << "  Normal end of execution.\n";

  std::cout << "\n";
  webbur::timestamp ( );
  
  return 0;
}
namespace webbur 
{
//****************************************************************************80

double parameter ( int dim, int offset )

//****************************************************************************80
//
//  Purpose:
//
//    PARAMETER is a user-supplied routine to retrieve parameters.
//
//  Discussion:
//
//    The declaration for this function is in SANDIA_RULES.H
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license. 
//
//  Modified:
//
//    02 April 2010
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, int DIM, the spatial dimension.
//
//    Input, int OFFSET, the offset of the parameter within the 
//    spatial dimension.
//
//    Output, double PARAMETER, the value of the OFFSET-th parameter
//    associated with the DIM-th dimension.
//
{
  int i;
  int j;
  double value;

  j = 0;
  for ( i = 0; i < dim; i++ )
  {
    j = j + NP[i];
  }
  value = P[j+offset];

  return value;
}
}
//****************************************************************************80

void sgmg_unique_index_tests ( double tol )

//****************************************************************************80
//
//  Purpose:
//
//    SGMG_UNIQUE_INDEX_TESTS calls SGMG_UNIQUE_INDEX_TEST.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license.
//
//  Modified:
//
//    01 January 2011
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, double TOL, a tolerance for point equality.
//    A value of sqrt ( eps ) is reasonable, and will allow the code to
//    consolidate points which are equal, or very nearly so.  A value of
//    -1.0, on the other hand, will force the code to use every point, 
//    regardless of duplication.
//
{
  int dim_num;
  int growth;
  GWPointer2 *gw_compute_order;
  GWPointer *gw_compute_points;
  int level_max_max;
  int level_max_min;
  int np_sum;

  std::cout << "\n";
  std::cout << "SGMG_UNIQUE_INDEX_TESTS\n";
  std::cout << "  Call SGMG_UNIQUE_INDEX_TEST with various arguments.\n";
  std::cout << "\n";
  std::cout << "  All tests will use a point equality tolerance of " << tol << "\n";

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 2;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::clenshaw_curtis_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_exp_cc;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 2;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::patterson_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_exp_gp;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::legendre_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_wn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::laguerre_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_nn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::gen_laguerre_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_nn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 1;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  P[0] = 1.5;
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::fejer2_points;
  gw_compute_points[1] = webbur::jacobi_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_f2;
  gw_compute_order[1] = webbur::level_to_order_linear_nn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 2;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  P[0] = 0.5;
  P[1] = 1.5;
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 2;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::gen_hermite_points;
  gw_compute_points[1] = webbur::hermite_genz_keister_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_linear_wn;
  gw_compute_order[1] = webbur::level_to_order_exp_hgk;
  NP = new int[dim_num];
  NP[0] = 1;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  P[0] = 2.0;
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  dim_num = 3;
  level_max_min = 0;
  level_max_max = 2;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::clenshaw_curtis_points;
  gw_compute_points[1] = webbur::legendre_points;
  gw_compute_points[2] = webbur::hermite_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_cc;
  gw_compute_order[1] = webbur::level_to_order_linear_wn;
  gw_compute_order[2] = webbur::level_to_order_linear_wn;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  NP[2] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;
//
//  Dimension 2, Level 4, Rule 3, slow exponential growth.
//
  dim_num = 2;
  level_max_min = 4;
  level_max_max = 4;
  growth = 0;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::patterson_points;
  gw_compute_points[1] = webbur::patterson_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_gp;
  gw_compute_order[1] = webbur::level_to_order_exp_gp;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;
//
//  Dimension 2, Level 4, Rule 3, moderate exponential growth.
//
  dim_num = 2;
  level_max_min = 4;
  level_max_max = 4;
  growth = 1;
  gw_compute_points = new GWPointer[dim_num];
  gw_compute_points[0] = webbur::patterson_points;
  gw_compute_points[1] = webbur::patterson_points;
  gw_compute_order = new GWPointer2[dim_num];
  gw_compute_order[0] = webbur::level_to_order_exp_gp;
  gw_compute_order[1] = webbur::level_to_order_exp_gp;
  NP = new int[dim_num];
  NP[0] = 0;
  NP[1] = 0;
  np_sum = webbur::i4vec_sum ( dim_num, NP );
  P = new double[np_sum];
  sgmg_unique_index_test ( dim_num, level_max_min, level_max_max, growth, 
    gw_compute_points, tol, gw_compute_order );
  delete [] gw_compute_order;
  delete [] gw_compute_points;
  delete [] NP;
  delete [] P;

  return;
}
//***************************************************************************80

void sgmg_unique_index_test 
(
  int dim_num,
  int level_max_min, 
  int level_max_max,
  int growth,  
  void ( *gw_compute_points[] ) ( int order, int dim, double w[] ),
  double tol,
  int ( *gw_compute_order[] ) ( int level, int growth )
)

//***************************************************************************80
//
//  Purpose:
//
//    SGMG_UNIQUE_INDEX_TEST tests SGMG_UNIQUE_INDEX.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license.
//
//  Modified:
//
//    01 January 2011
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, int DIM_NUM, the spatial dimension.
//
//    Input, int LEVEL_MAX_MIN, LEVEL_MAX_MAX, the minimum and
//    maximum values of LEVEL_MAX.
//
//    Input, int GROWTH, the growth rule. 
//    0, slow;
//    1, moderate;
//    2, full.
//
//    Input, void ( *GW_COMPUTE_POINTS[] ) ( int order, int dim, double x[] ),
//    an array of pointers to functions which return the 1D quadrature points 
//    associated with each spatial dimension.
//
//    Input, double TOL, a tolerance for point equality.
//
//    Input, int ( *GW_COMPUTE_ORDER[] ) ( int level, int growth ),
//    an array of pointers to functions which return the order of the
//    1D quadrature rule of a given level and growth rule.
//
{
  int dim;
  int level_max;
  int level_min;
  int point;
  int point_num;
  int point_total_num;
  int *sparse_unique_index;

  std::cout << "\n";
  std::cout << "SGMG_UNIQUE_INDEX_TEST\n";
  std::cout << "  SGMG_UNIQUE_INDEX returns a mapping between\n";
  std::cout << "  the nonunique and unique points in a sparse grid.\n";

  for ( level_max = level_max_min; level_max <= level_max_max; level_max++ )
  {
    point_total_num = webbur::sgmg_size_total ( dim_num, level_max, growth,
      gw_compute_order );

    point_num = webbur::sgmg_size ( dim_num, level_max, gw_compute_points, tol, 
      growth, gw_compute_order );

    std::cout << "\n";
    std::cout << " LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM\n";
    std::cout << "                        Unique     Total\n";

    level_min = webbur::i4_max ( 0, level_max + 1 - dim_num );

    std::cout << "\n";
    std::cout << "  " << std::setw(8) << level_min
              << "  " << std::setw(8) << level_max
              << "  " << std::setw(8) << point_num
              << "  " << std::setw(8) << point_total_num << "\n";

    sparse_unique_index = new int[point_total_num];

    webbur::sgmg_unique_index ( dim_num, level_max, gw_compute_points, tol, 
      point_num, point_total_num, growth, gw_compute_order, sparse_unique_index );

    std::cout << "\n";
    std::cout << "     POINT    UNIQUE\n";
    std::cout << "\n";
    for ( point = 0; point < point_total_num; point++ )
    {
      std::cout << "  " << std::setw(8) << point
                << "  " << std::setw(8) << sparse_unique_index[point] << "\n";
    }
    delete [] sparse_unique_index;
  }
  return;
}

