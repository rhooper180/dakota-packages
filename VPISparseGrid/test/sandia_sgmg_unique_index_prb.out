13 January 2012 10:40:48 AM

SGMG_UNIQUE_INDEX_PRB
  C++ version

SGMG_UNIQUE_INDEX_TESTS
  Call SGMG_UNIQUE_INDEX_TEST with various arguments.

  All tests will use a point equality tolerance of 1.49012e-08

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1         5         7

     POINT    UNIQUE

         0         2
         1         0
         2         2
         3         4
         4         1
         5         2
         6         3

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         1         2        13        25

     POINT    UNIQUE

         0         1
         1         6
         2        11
         3         4
         4         6
         5         8
         6         1
         7         3
         8         6
         9         9
        10        11
        11         0
        12         4
        13        10
        14         1
        15         6
        16        11
        17         2
        18         8
        19        12
        20         4
        21         5
        22         6
        23         7
        24         8

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1         5         7

     POINT    UNIQUE

         0         2
         1         0
         2         2
         3         4
         4         1
         5         2
         6         3

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         1         2        15        27

     POINT    UNIQUE

         0         1
         1         7
         2        13
         3         5
         4         7
         5         9
         6         1
         7         3
         8         7
         9        11
        10        13
        11         0
        12         5
        13        12
        14         1
        15         7
        16        13
        17         2
        18         9
        19        14
        20         4
        21         5
        22         6
        23         7
        24         8
        25         9
        26        10

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1         7         9

     POINT    UNIQUE

         0         3
         1         0
         2         1
         3         3
         4         5
         5         6
         6         2
         7         3
         8         4

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         1         2        23        37

     POINT    UNIQUE

         0         1
         1         5
         2        11
         3        17
         4        21
         5         9
         6        11
         7        13
         8         1
         9         3
        10         5
        11         7
        12        11
        13        15
        14        17
        15        19
        16        21
        17         0
        18         4
        19         9
        20        16
        21        20
        22         1
        23         5
        24        11
        25        17
        26        21
        27         2
        28         6
        29        13
        30        18
        31        22
        32         8
        33        10
        34        11
        35        12
        36        14

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1         8         9

     POINT    UNIQUE

         0         3
         1         0
         2         1
         3         3
         4         6
         5         7
         6         2
         7         4
         8         5

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         1         2        29        37

     POINT    UNIQUE

         0         1
         1         6
         2        12
         3        21
         4        26
         5        11
         6        14
         7        16
         8         1
         9         4
        10         6
        11         9
        12        12
        13        19
        14        21
        15        24
        16        26
        17         0
        18         5
        19        11
        20        20
        21        25
        22         2
        23         7
        24        14
        25        22
        26        27
        27         3
        28         8
        29        16
        30        23
        31        28
        32        10
        33        13
        34        15
        35        17
        36        18

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1         8         9

     POINT    UNIQUE

         0         3
         1         0
         2         1
         3         3
         4         6
         5         7
         6         2
         7         4
         8         5

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         1         2        29        37

     POINT    UNIQUE

         0         1
         1         6
         2        13
         3        21
         4        26
         5        11
         6        14
         7        16
         8         1
         9         4
        10         6
        11         9
        12        13
        13        19
        14        21
        15        24
        16        26
        17         0
        18         5
        19        11
        20        20
        21        25
        22         2
        23         7
        24        14
        25        22
        26        27
        27         3
        28         8
        29        16
        30        23
        31        28
        32        10
        33        12
        34        15
        35        17
        36        18

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1        10        11

     POINT    UNIQUE

         0         5
         1         0
         2         1
         3         2
         4         5
         5         7
         6         8
         7         9
         8         3
         9         4
        10         6

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         1         2        41        51

     POINT    UNIQUE

         0         3
         1         8
         2        13
         3        21
         4        28
         5        33
         6        38
         7        17
         8        20
         9        23
        10         0
        11         3
        12         5
        13         8
        14        10
        15        13
        16        15
        17        21
        18        25
        19        28
        20        30
        21        33
        22        35
        23        38
        24        40
        25         1
        26         6
        27        11
        28        17
        29        26
        30        31
        31        36
        32         2
        33         7
        34        12
        35        20
        36        27
        37        32
        38        37
        39         4
        40         9
        41        14
        42        23
        43        29
        44        34
        45        39
        46        16
        47        18
        48        19
        49        22
        50        24

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1         5         7

     POINT    UNIQUE

         0         2
         1         0
         2         2
         3         4
         4         1
         5         2
         6         3

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         1         2        19        29

     POINT    UNIQUE

         0         2
         1        13
         2        16
         3        12
         4        13
         5         9
         6         0
         7         4
         8        13
         9        14
        10        18
        11         1
        12        12
        13        15
        14         2
        15        13
        16        16
        17         3
        18         9
        19        17
        20         5
        21         6
        22        12
        23         7
        24        13
        25         8
        26         9
        27        10
        28        11

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         0         1         1

     POINT    UNIQUE

         0         0

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         1         9        12

     POINT    UNIQUE

         0         4
         1         0
         2         1
         3         4
         4         7
         5         8
         6         2
         7         4
         8         6
         9         3
        10         4
        11         5

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         0         2        41        70

     POINT    UNIQUE

         0        23
         1         3
         2         8
         3        23
         4        32
         5        39
         6        14
         7        23
         8        26
         9        22
        10        23
        11        20
        12         3
        13         5
        14         8
        15        11
        16        23
        17        29
        18        32
        19        35
        20        39
        21         0
        22         6
        23        14
        24        30
        25        36
        26         3
        27         8
        28        23
        29        32
        30        39
        31         4
        32        10
        33        26
        34        34
        35        40
        36        12
        37        16
        38        23
        39        24
        40        28
        41         1
        42         7
        43        22
        44        31
        45        37
        46         3
        47         8
        48        23
        49        32
        50        39
        51         2
        52         9
        53        20
        54        33
        55        38
        56        13
        57        22
        58        25
        59        14
        60        23
        61        26
        62        15
        63        20
        64        27
        65        17
        66        18
        67        23
        68        19
        69        21

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         3         4        33        97

     POINT    UNIQUE

         0         1
         1         6
         2        11
         3        16
         4        21
         5        26
         6        31
         7         4
         8        14
         9        24
        10         6
        11        16
        12        26
        13         8
        14        18
        15        28
        16         4
        17        14
        18        24
        19         6
        20        16
        21        26
        22         8
        23        18
        24        28
        25        13
        26        14
        27        15
        28        16
        29        17
        30        18
        31        19
        32         1
        33         6
        34        11
        35        16
        36        21
        37        26
        38        31
        39         0
        40         4
        41        10
        42        14
        43        20
        44        24
        45        30
        46         1
        47         6
        48        11
        49        16
        50        21
        51        26
        52        31
        53         2
        54         8
        55        12
        56        18
        57        22
        58        28
        59        32
        60         4
        61        14
        62        24
        63         6
        64        16
        65        26
        66         8
        67        18
        68        28
        69         3
        70        13
        71        23
        72         4
        73        14
        74        24
        75         5
        76        15
        77        25
        78         6
        79        16
        80        26
        81         7
        82        17
        83        27
        84         8
        85        18
        86        28
        87         9
        88        19
        89        29
        90        13
        91        14
        92        15
        93        16
        94        17
        95        18
        96        19

SGMG_UNIQUE_INDEX_TEST
  SGMG_UNIQUE_INDEX returns a mapping between
  the nonunique and unique points in a sparse grid.

 LEVEL_MIN LEVEL_MAX POINT_NUM POINT_NUM
                        Unique     Total

         3         4        97       241

     POINT    UNIQUE

         0         1
         1         6
         2        11
         3        20
         4        29
         5        34
         6        39
         7        48
         8        57
         9        62
        10        67
        11        76
        12        85
        13        90
        14        95
        15         4
        16        16
        17        32
        18        44
        19        60
        20        72
        21        88
        22         6
        23        20
        24        34
        25        48
        26        62
        27        76
        28        90
        29         8
        30        24
        31        36
        32        52
        33        64
        34        80
        35        92
        36        14
        37        42
        38        70
        39        16
        40        44
        41        72
        42        18
        43        46
        44        74
        45        20
        46        48
        47        76
        48        22
        49        50
        50        78
        51        24
        52        52
        53        80
        54        26
        55        54
        56        82
        57        41
        58        42
        59        43
        60        44
        61        45
        62        46
        63        47
        64        48
        65        49
        66        50
        67        51
        68        52
        69        53
        70        54
        71        55
        72         1
        73         6
        74        11
        75        20
        76        29
        77        34
        78        39
        79        48
        80        57
        81        62
        82        67
        83        76
        84        85
        85        90
        86        95
        87         0
        88         4
        89        10
        90        16
        91        28
        92        32
        93        38
        94        44
        95        56
        96        60
        97        66
        98        72
        99        84
       100        88
       101        94
       102         1
       103         6
       104        11
       105        20
       106        29
       107        34
       108        39
       109        48
       110        57
       111        62
       112        67
       113        76
       114        85
       115        90
       116        95
       117         2
       118         8
       119        12
       120        24
       121        30
       122        36
       123        40
       124        52
       125        58
       126        64
       127        68
       128        80
       129        86
       130        92
       131        96
       132         3
       133        14
       134        31
       135        42
       136        59
       137        70
       138        87
       139         4
       140        16
       141        32
       142        44
       143        60
       144        72
       145        88
       146         5
       147        18
       148        33
       149        46
       150        61
       151        74
       152        89
       153         6
       154        20
       155        34
       156        48
       157        62
       158        76
       159        90
       160         7
       161        22
       162        35
       163        50
       164        63
       165        78
       166        91
       167         8
       168        24
       169        36
       170        52
       171        64
       172        80
       173        92
       174         9
       175        26
       176        37
       177        54
       178        65
       179        82
       180        93
       181        13
       182        41
       183        69
       184        14
       185        42
       186        70
       187        15
       188        43
       189        71
       190        16
       191        44
       192        72
       193        17
       194        45
       195        73
       196        18
       197        46
       198        74
       199        19
       200        47
       201        75
       202        20
       203        48
       204        76
       205        21
       206        49
       207        77
       208        22
       209        50
       210        78
       211        23
       212        51
       213        79
       214        24
       215        52
       216        80
       217        25
       218        53
       219        81
       220        26
       221        54
       222        82
       223        27
       224        55
       225        83
       226        41
       227        42
       228        43
       229        44
       230        45
       231        46
       232        47
       233        48
       234        49
       235        50
       236        51
       237        52
       238        53
       239        54
       240        55

SGMG_UNIQUE_INDEX_PRB
  Normal end of execution.

13 January 2012 10:40:48 AM
